### ISSUE TITLE
[Feature]: Context Caching for Vertex AI

### ISSUE THREAD

**Issue by DreamGenX** [BerriAI/litellm#6898]
### The Feature

It looks like Gemini context caching does not work when using Vertex AI.
Doing a cursory search, it looks like this part would need to be implemented to support Vertex AI: https://github.com/BerriAI/litellm/blob/c73ce95c01d33eeebe8adb4df2b899053261e405/litellm/llms/vertex_ai_and_google_ai_studio/gemini/transformation.py#L392

### Motivation, pitch

Context cachin is already supported using the Gemini API and it's a good way to reduce costs.

**Comment by codenprogressive** [comment]
+1
Any updates on enabling context caching for LLMs (Gemini, Claude, etc) available through VertexAI?

**Comment by ghost** [comment]
@codenprogressive just tested claude on vertex ai - it doesn't look like prompt caching is available there yet. 

I plan on picking up vertex ai context caching this week

**Comment by codenprogressive** [comment]
Hey @krrishdholakia, indeed Claude on Vertex still doesn't support prompt caching! (there is a feature request: https://github.com/anthropics/anthropic-sdk-python/issues/653)

I think for now we can enable it for **Gemini on VertexAI**.

**Comment by emerzon** [comment]
AWS has started a preview for the prompt caching for Claude: https://pages.awscloud.com/promptcaching-Preview.html
Hope Vertex comes up next

**Comment by emerzon** [comment]
Vertex support is now live: https://cloud.google.com/vertex-ai/generative-ai/docs/partner-models/claude-prompt-caching#use_prompt_caching

**Comment by ghost** [comment]
@emerzon i believe this ticket is for vertex ai gemini, the vertex ai anthropic prompt caching issue is probably separate. i'm working on it now though, since we shouldn't be passing any extra headers to it.

**Comment by sammcj** [comment]
This is absolutely critical for things like agentic coding, where prompt caching results in around a 98% saving in cost.

Prompt caching works _really_ well with GCP VertexAI directly, so this will be a bit of a blocker for a lot of folks wanting to use LiteLLM, or using LiteLLM and wondering why they're paying 98% more.

**Comment by sammcj** [comment]
The code is a little hard to follow but it almost looks like there's just one header missing, if that is all that's required here is a PR: https://github.com/BerriAI/litellm/pull/9605

**Comment by ghost** [comment]
Merged https://github.com/BerriAI/litellm/pull/9605

Anthropic Prompt caching on vertex ai is now on main

**Comment by ghost** [comment]
> The code is a little hard to follow

hey @sammcj how could this have been easier?

**Comment by mollux** [comment]
@krrishdholakia 
This issue should be reopened, as https://github.com/BerriAI/litellm/pull/9605 was reverted in https://github.com/BerriAI/litellm/pull/9670

Did anyone got Anthropic Prompt caching working for vertex AI?

**Comment by mollux** [comment]
@sammcj did the patch you made fixed the caching issue for Claude models hosted on vertex AI? I applied the patch but it's not working

**Comment by sammcj** [comment]
@mollux - I thought I had it working but no go sorry :(

**Comment by mio-inamijima** [comment]
Hello! After reading through the thread, I understand that context caching for Vertex AI Gemini has been implemented and is working, while for Claude it's still not completed. I'd be willing to help document the context caching functionality for Vertex AI. Should I wait until the Claude implementation is complete, or would it be helpful to document the Gemini part now?

**Comment by nang-dev** [comment]
@mio-inamijima From my understanding, it's the other way around. Claude has been implemented and is working, while for  Vertex AI Gemini it's still not completed.

**Comment by mio-inamijima** [comment]
@nang-dev 

Thank you, I appreciate your comment.

But is that true?

> [@sammcj](https://github.com/sammcj) did the patch you made fixed the caching issue for Claude models hosted on vertex AI? I applied the patch but it's not working

> [@mollux](https://github.com/mollux) - I thought I had it working but no go sorry :(

From these replies, I thought that context caching for Claude on Vertexai is not working yet.

**Comment by mio-inamijima** [comment]
I just checked out if conext/prompt caching on gemini/claude on vertex ai will work, and on 1.67.4, it didn't.
Release note of 1.67.5 says nothing about it, so I guess that this case is sitll open for both gemini and claude
thank you.

**Comment by nang-dev** [comment]
I'm seeing the same, and am hoping for prompt caching capabilities on vertex ai + gemini

**Comment by sammcj** [comment]
Any update on getting prompt caching with VertexAI working in LiteLLM?

**Comment by krrishdholakia** [comment]
Vertex AI Anthropic prompt caching is **already** working with litellm proxy. 

Just tested with this script - 

<img width="616" height="94" alt="Image" src="https://github.com/user-attachments/assets/86fd663b-09a7-4fef-a928-5d5285df2832" />

```python
"""Minimal reproduction of litellm tool message cache_control bug."""

import asyncio
import os
import openai

# Initialize OpenAI client for proxy
openai_client = openai.OpenAI(
    api_key="sk-1234",
    base_url="http://0.0.0.0:4000"
)

LITELLM_MODEL = "vertex_ai/claude-3-5-sonnet@20240620"


def test_openai_proxy():
    print("\n--- Testing with OpenAI client via litellm proxy ---")
    # Same tool message with cache_control
    tool_content = "Result: 4. " * 1000  # ~10k chars

    messages = [
        {"role": "user", "content": "Calculate 2+2"},
        {
            "role": "assistant",
            "content": None,
            "tool_calls": [
                {
                    "id": "call_proxy_123",
                    "type": "function",
                    "function": {"name": "calc", "arguments": "{}"},
                }
            ],
        },
        {
            "role": "tool",
            "tool_call_id": "call_proxy_123",
            "content": [
                {
                    "type": "text",
                    "text": tool_content,
                    "cache_control": {"type": "ephemeral"},
                }
            ],
        },
    ]

    tools = [
        {
            "type": "function",
            "function": {
                "name": "calc",
                "description": "Calculator",
                "parameters": {"type": "object", "properties": {}},
            },
        }
    ]

    # Make two identical requests - second should use cache
    for i in range(2):
        response = openai_client.chat.completions.create(
            model=LITELLM_MODEL,
            messages=messages + [{"role": "user", "content": f"What's {i+3}+{i+3}?"}],
            tools=tools,
            max_tokens=50,
        )
        usage = response.usage
        cache_write = getattr(usage, "cache_creation_input_tokens", "N/A")
        cache_read = getattr(usage, "cache_read_input_tokens", "N/A")
        print(
            f"Proxy Request {i+1}: cache_write={cache_write}, cache_read={cache_read}"
        )


async def main():
    print("--- Testing with litellm ---")
    test_openai_proxy()


if __name__ == "__main__":
    asyncio.run(main())
```


the only thing missing is tool call cache control, which is added now via - https://github.com/BerriAI/litellm/pull/12668

**Comment by sammcj** [comment]
FYI the litellm docs (https://docs.litellm.ai/docs/completion/prompt_caching) still don't list Vertex caching as being available.

**Comment by cangozpi** [comment]
It seems like the Vertex AI caching support is added:
```bash
>>> from litellm.utils import supports_prompt_caching
>>> supports_prompt_caching(model="vertex_ai/gemini-2.5-flash-lite")
True
```

but I cannot figure out how to use caching as I cannot find any cache related options in the docs(https://docs.litellm.ai/docs/completion/input#input-params-1) or in the following output:

```bash
>>> from litellm import get_supported_openai_params
>>> get_supported_openai_params(model="gemini-2.5-flash-lite", custom_llm_provider="vertex_ai")
['temperature', 'top_p', 'max_tokens', 'max_completion_tokens', 'stream', 'tools', 'functions', 'tool_choice', 'response_format', 'n', 'stop', 'frequency_penalty', 'presence_penalty', 'extra_headers', 'seed', 'logprobs', 'top_logprobs', 'modalities', 'parallel_tool_calls', 'web_search_options', 'reasoning_effort', 'thinking']
````

I would appreciate any help on this.

**Comment by anthrotype** [comment]
> FYI LiteLLM's Vertex caching with Gemini models still doesn't actually work

The context caching endpoint is only used when `gemini_api_key is not None`, see:

https://github.com/BerriAI/litellm/blob/9a62b9bdb9ff217a0683047756588b2f5bd59c27/litellm/llms/vertex_ai/gemini/transformation.py#L490-L510

`gemini_api_key` is only set when the `custom_llm_provider` is "gemini", not when it is "vertex_ai"; see around these line:

https://github.com/BerriAI/litellm/blob/9a62b9bdb9ff217a0683047756588b2f5bd59c27/litellm/main.py#L2802

(The naming of these modules is super confusing by the way).
